native/src/main/java/io/ballerina/stdlib/data/csvdata/utils/CsvUtils.java

```diff
@@ -0,0 +1,350 @@
+        }
+        return ValueUtils.convert(csv, targetType);
+    }
+
+    public static void checkRequiredFieldsAndLogError(Map<String, Field> filedHierarchy, boolean absentAsNilableType) {
+        filedHierarchy.values().forEach(field -> {
+            if (absentAsNilableType && field.getFieldType().isNilable()) {
+                return;
+            }
+            if (SymbolFlags.isFlagOn(field.getFlags(), SymbolFlags.REQUIRED)) {
+                throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_FIELD_IN_CSV, field.getFieldName());
+            }
+        });
+    }
+
+    public static boolean isHeaderFieldsEmpty(Map<String, Field> currentField) {
+        for (Field field: currentField.values()) {
+            if (SymbolFlags.isFlagOn(field.getFlags(), SymbolFlags.REQUIRED)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public static boolean checkTypeCompatibility(Type constraintType, Object csv, boolean stringConversion) {
+        int tag = constraintType.getTag();
+        if ((csv instanceof BString && (stringConversion || tag == TypeTags.STRING_TAG
+                || tag == TypeTags.CHAR_STRING_TAG || isJsonOrAnyDataOrAny(tag)))
+                || (csv instanceof Long && (tag == TypeTags.INT_TAG
+                || tag == TypeTags.FLOAT_TAG || tag == TypeTags.DECIMAL_TAG || tag == TypeTags.BYTE_TAG
+                || tag == TypeTags.SIGNED8_INT_TAG || tag == TypeTags.SIGNED16_INT_TAG
+                || tag == TypeTags.SIGNED32_INT_TAG || tag == TypeTags.UNSIGNED8_INT_TAG
+                || tag == TypeTags.UNSIGNED16_INT_TAG || tag == TypeTags.UNSIGNED32_INT_TAG
+                || isJsonOrAnyDataOrAny(tag)))
+                || (csv instanceof BDecimal && ((tag == TypeTags.DECIMAL_TAG
+                || tag == TypeTags.FLOAT_TAG || tag == TypeTags.INT_TAG) || isJsonOrAnyDataOrAny(tag)))
+                || (csv instanceof Double && ((tag == TypeTags.FLOAT_TAG
+                || tag == TypeTags.DECIMAL_TAG || tag == TypeTags.INT_TAG) || isJsonOrAnyDataOrAny(tag)))
+                || (Boolean.class.isInstance(csv) && (tag == TypeTags.BOOLEAN_TAG || isJsonOrAnyDataOrAny(tag)))
+                || (csv == null && (tag == TypeTags.NULL_TAG || isJsonOrAnyDataOrAny(tag)))) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private static boolean isJsonOrAnyDataOrAny(int tag) {
+        return tag == TypeTags.JSON_TAG || tag == TypeTags.ANYDATA_TAG || tag == TypeTags.ANY_TAG;
+    }
+
+    public static int getTheActualExpectedType(Type type) {
+        if (type instanceof TupleType) {
+            TupleType tupleType = (TupleType) type;
+            if (tupleType.getRestType() != null) {
+                return -1;
+            }
+            return tupleType.getTupleTypes().size();
+        } else {
+            return ((ArrayType) type).getSize();
+        }
+    }
+
+    public static HashMap<String, String>
```

Comment:
Do the callers depend on the Hashmap property? otherwise I think it is better to return a `Map<String, String>